shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/data/remote/apiclient/core/ApiClient.kt

```diff
@@ -0,0 +1,109 @@
+import co.nimblehq.ic.kmm.suv.data.remote.model.TokenApiModel
+import co.nimblehq.ic.kmm.suv.domain.model.AppError
+import co.nimblehq.jsonapi.json.JsonApi
+import co.nimblehq.jsonapi.model.JsonApiException
+import io.github.aakira.napier.DebugAntilog
+import io.github.aakira.napier.LogLevel
+import io.github.aakira.napier.Napier
+import io.ktor.client.*
+import io.ktor.client.engine.*
+import io.ktor.client.plugins.auth.*
+import io.ktor.client.plugins.auth.providers.*
+import io.ktor.client.plugins.contentnegotiation.*
+import io.ktor.client.plugins.logging.*
+import io.ktor.client.plugins.logging.LogLevel.ALL
+import io.ktor.client.request.*
+import io.ktor.client.statement.*
+import io.ktor.http.*
+import io.ktor.serialization.kotlinx.json.*
+import kotlinx.coroutines.flow.*
+import kotlinx.serialization.json.Json
+
+class ApiClient(
+    engine: HttpClientEngine,
+    tokenLocalDataSource: TokenLocalDataSource? = null,
+    tokenRemoteDataSource: TokenRemoteDataSource? = null
+) {
+
+    val httpClient: HttpClient
+    val json = Json {
+        prettyPrint = true
+        isLenient = true
+        encodeDefaults = true
+        ignoreUnknownKeys = true
+    }
+
+    private val tokenLocalDataSource: TokenLocalDataSource?
+
+    init {
+        this.tokenLocalDataSource = tokenLocalDataSource
+        Napier.takeLogarithm()
+        Napier.base(DebugAntilog())
+        httpClient = HttpClient(engine = engine) {
+            install(Logging) {
+                level = ALL
+                logger = object : Logger {
+                    override fun log(message: String) {
+                        Napier.log(LogLevel.DEBUG, message = message)
+                    }
+                }
+            }
+
+            install(ContentNegotiation) {
+                json(json)
+            }
+
+            if (tokenLocalDataSource != null && tokenRemoteDataSource != null) {
+                install(Auth) {
+                    bearer {
+                        loadTokens {
+                            tokenLocalDataSource.getToken()?.run {
+                                BearerTokens(accessToken, refreshToken)
+                            }
+                        }
+
+                        sendWithoutRequest { request ->
+                            request.url.host == Url(BuildKonfig.BASE_URL).host
+                        }
+
+                        refreshTokens {
+                            val token = tokenRemoteDataSource
+                                .refreshToken(RefreshTokenApiBody(
+                                    refreshToken = oldTokens?.refreshToken ?: "")
+                                )
+                                .last()
+                            tokenLocalDataSource.save(token)
+                            tokenLocalDataSource.getToken()?.run {
+                                BearerTokens(accessToken, refreshToken)
+                            }
```

Comment:
Should we use the new `token` to build `BearerTokens` directly? Instead of getting the latest value that we just saved in the previous step?

```suggestion
                            BearerTokens(token.accessToken, token.refreshToken)
```

Reply:
@luongvo This is a good point! I did return the token directly. However, after checking [this step](https://ktor.io/docs/bearer-client.html#step10), I decided to follow the example. I think it's better to make sure we apply the same token that is gotten from the local resource.

What do you think? If you prefer to return the value directly, I'm good with that as well :D

Reply:
@suho In their case, that makes sense because they saved and load the whole `BearerTokens` model, so they don't need to build it again. However here, it seems we need to save a TokenModel, load it back and get its fields out for building `BearerTokens`. It has a bit of difference 🤷‍♂️

Reply:
@luongvo Let me update with returning the token directly approach 🙈

Reply:
Fixed in https://github.com/suho/kmm-ic/pull/94/commits/a15e1726f30b435a60f468632e94599388033dbb